### PR TITLE
Add a new getting started guide

### PR DESCRIPTION
**Pull request by sgrif**
[Rendered](https://github.com/sgrif/diesel/blob/sg-new-guide/guide.md)

This is not actually meant for this repo, this will end up on the new
website. However, I wanted to write out all the content in markdown and
review it here before I go through all the formatting, etc over there.

After the website is live, the README will get slimmed down by quite a
lot, mostly pointing to this guide. One thing that I couldn't figure out
how to fit in naturally was passing a struct to `update().set`, or using
methods like `save_changes`. Maybe it's fine to skip them, as this isn't
intended to be an exhaustive guide?

I also want to follow this up with an "advanced" guide that covers joins
and associations in 0.6.

### CODE REVIEW

**Comment by sgrif** [comment]
Review? @samphippen @mfpiccolo @mcasper

**Comment by sgrif** [comment]
Also @derekprior @tessgriffin @steveklabnik will likely have feedback on this. Please rip it to shreds. <3

File: guide.md [hunk 1]
```diff
@@ -0,0 +1,388 @@
+Getting Started 2.0
+===
+
+For this guide, we're going to walk through some simple examples for each of the
+pieces of CRUD. Each step in this guide will build on the previous, and is meant
+to be followed along. This guide assumes that you're using PostgreSQL on
+nightly. We'll talk about how to use Diesel on stable in the final chapter.
```
Comment (derekprior) [inline guide.md]: "PostgreSQL and Rust nightly" or something similar? At first read I thought "nightly" applied to Postgres.

File: guide.md [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+Before we start, make sure you have PostgreSQL installed and running.
```
Comment (derekprior) [inline guide.md]: If this needs to be said then does it maybe deserve a link?

File: guide.md [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+
+The first thing we need to do is generate our project.
+
+```shell
+cargo new diesel_demo
+cd diesel_demo
+```
+
+We're going to use a [`.env`][dotenv-rust] to manage our environment variables
```
Comment (derekprior) [inline guide.md]: "a `.env` file..."

Comment (mcasper) [inline guide.md]: Do you pronounce it `dotenv` or `env`?

File: guide.md (hunk 2 above)
Comment (sgrif) [inline guide.md]: Do you think it needs to be said?

File: guide.md (hunk 3 above)
Comment (mcasper) [inline guide.md]: Ignore me. Misread the comment

Comment (sgrif) [inline guide.md]: "a dotenv file"

File: guide.md [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+for us, and [`diesel_cli`][diesel-cli] to help set up the database.
+
+[dotenv-rust]: https://github.com/slapresta/rust-dotenv
+[diesel-cli]: https://github.com/sgrif/diesel/diesel_cli
+
+```shell
+echo DATABASE_URL=postgres://localhost/diesel_demo > .env
+cargo install diesel_cli
```
Comment (mcasper) [inline guide.md]: The user also has to add `$CARGO_PATH/bin` to their PATH, worth mentioning? Or leave that to learning rust and cargo?

File: guide.md [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+diesel setup
+```
+
+This will create our database for us (if it didn't already exist), and create an
+empty migrations directory that we can use to manage our schema. More on that
+later. Let's add Diesel to our `Cargo.toml`. We'll also add `dotenv` so we can
```
Comment (derekprior) [inline guide.md]: The ordering of this all took me a bit by surprise because we did something related to `dotenv` already (albeit, we just added it's file) and only now are we declaring the dependency.

I gather that's possibly because `diesel_cli` depends on .env internally?

File: guide.md [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+    stdin().read_to_string(&mut body).unwrap();
+
+    let post = create_post(&connection, title, &body);
+    println!("\nSaved draft {} with id {}", title, post.id);
+}
+
+#[cfg(not(windows))]
+const EOF: &'static str = "CTRL+D";
+
+#[cfg(windows)]
```
Comment (sgrif) [inline guide.md]: Yes, I'm being "that guy"

File: guide.md [hunk 7]
```diff
@@ -0,0 +1,388 @@
...
+
+```
+Creating migrations/20160202154039_create_posts/up.sql
+Creating migrations/20160202154039_create_posts/down.sql
+```
+
+Every migration consists of a way to run it, and a way to undo it. `down.sql`
+should always leave your database in the state it was in before running `up.sql`
+when possible. We'll edit our migration to contain the following:
+
```
Comment (tessgriffin) [inline guide.md]: When you run the diesel migration, does it create just one file or two? Lines 62 and 63 imply 2 to me. If that's the case, you should say on line 68 "We'll edit our migrations".

File: guide.md (hunk 5 above)
Comment (sgrif) [inline guide.md]: Yes, `diesel_cli` depends directly on `dotenv`. Since it's a separate crate, there's no reason for it not to

Comment (derekprior) [inline guide.md]: Having read a bit more, I'd consider re-ordering everything after generating the project. Maybe:

1. set up your project dependencies in the `cargo.toml` file.
2. introduce diesel cli. What's it for? Why isn't it a project dependency? How do I install it?
3. Use .env and diesel cli to create our database.

File: guide.md [hunk 8]
```diff
@@ -0,0 +1,388 @@
...
+```shell
+cargo new diesel_demo
+cd diesel_demo
+```
+
+We're going to use a [`.env`][dotenv-rust] to manage our environment variables
+for us, and [`diesel_cli`][diesel-cli] to help set up the database.
+
+[dotenv-rust]: https://github.com/slapresta/rust-dotenv
+[diesel-cli]: https://github.com/sgrif/diesel/diesel_cli
```
Comment (derekprior) [inline guide.md]: This link does not work

File: guide.md (hunk 7 above)
Comment (sgrif) [inline guide.md]: It creates two files. However, I think that `up.sql` and `down.sql` can be collectively referred to as "a migration". 2 files != 2 migrations.

File: guide.md (hunk 4 above)
Comment (sgrif) [inline guide.md]: I opted to leave it out, as cargo outputs this for me:

```
be sure to add `/Users/sean/.multirust/toolchains/nightly/cargo/bin` to your PATH to be able to run the installed binaries
```

I'm fine with calling it out if you think we should, though.

File: guide.md [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+```toml
+[dependencies]
+diesel = "^0.5.0"
+diesel_codegen = { version = "^0.5.0", default-features = false, features = ["nightly", "postgres"] }
+dotenv = "^0.8.0"
+dotenv_macros = "^0.8.0"
+```
+
+We'll also add this to the top of `src/lib.rs`, which will allow us to use
+features from `diesel_codegen`.
```
Comment (derekprior) [inline guide.md]: Can you give a quick summary of what `diesel_codegen` does or at least let me know we're going to cover that later if indeed we do?

Comment (sgrif) [inline guide.md]: I'm open to suggestions on this one. codegen being a separate crate might be fairly alien to devs from other languages. I don't think I ever really cover what it does, and I don't have a great "quick summary" other than "it generates code".

Comment (sgrif) [inline guide.md]: I could maybe say something like "which gives us several useful class annotations, as well as the `infer_schema!` macro. We'll see both of those soon."

File: guide.md [hunk 10]
```diff
@@ -0,0 +1,388 @@
...
+```
+
+You should see output which looks something like this:
+
+```
+Creating migrations/20160202154039_create_posts/up.sql
+Creating migrations/20160202154039_create_posts/down.sql
+```
+
+Every migration consists of a way to run it, and a way to undo it. `down.sql`
```
Comment (derekprior) [inline guide.md]: Maybe something like: "Migrations allow us to evolve the database schema over time. Each migration can be applied (`up.sql`) or rolled-back (`down.sql`). Applying and immediately rolling-back a migration should leave your database schema unchanged."

My chief worry was the term `run`. You `run` your down migration too, really...

Comment (sgrif) [inline guide.md]: Good point. I picked `run` because that's the CLI command, but I like your description better (though we say `revert` instead of `rollback`)

File: guide.md (hunk 7 above)
Comment (tessgriffin) [inline guide.md]: I think it would make sense then to visually separate the two. Even just a white line. Something to indicate that there are two files (even though yes, it's one migration)

File: guide.md [hunk 11]
```diff
@@ -0,0 +1,388 @@
...
+  published BOOLEAN NOT NULL DEFAULT 'f'
+)
+
+-- down.sql
+DROP TABLE posts
+```
+
+We can run our new migration with `diesel migration run`. It's usually a good
+idea to make sure that `down.sql` is correct. You can do a quick sanity check by
+doing `diesel migration redo`.
```
Comment (derekprior) [inline guide.md]: which will roll back and then apply your migration again.

File: guide.md [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+  title VARCHAR NOT NULL,
+  body TEXT NOT NULL,
+  published BOOLEAN NOT NULL DEFAULT 'f'
+)
+
+-- down.sql
+DROP TABLE posts
+```
+
+We can run our new migration with `diesel migration run`. It's usually a good
```
Comment (derekprior) [inline guide.md]: `run` troubles me here for the previously specified reason. Maybe I'm overthinking it.

File: guide.md [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+database connection.
+
+```rust
+// src/lib.rs
+#[macro_use]
+extern crate diesel;
+extern crate dotenv;
+
+use diesel::prelude::*;
+use diesel::pg::PgConnection;
```
Comment (mcasper) [inline guide.md]: A few things here, most of which I assume were intentional and will be true in 0.5
* PgConnection lives in `diesel::connection::pg::PgConnection` right now
* "postgres" isn't currently a valid feature (listed above in the Cargo.toml)
* `diesel::Connection` has to be imported for the `Connection.establish` to be available

Again, I assume most of these things will be true by 0.5, just bringing them up

File: guide.md (hunk 7 above)
Comment (sgrif) [inline guide.md]: Yes, they will be in separate panes on the website.

File: guide.md (hunk 13 above)
Comment (sgrif) [inline guide.md]: This code all runs when used on the code as of https://github.com/sgrif/diesel/pull/175 (which I'm still waiting for a review on. :P)

File: guide.md (hunk 12 above)
Comment (sgrif) [inline guide.md]: I don't have strong feelings about it, though `apply` feels weird to me. @mfpiccolo @mcasper do either of you have an opinion here?

Comment (sgrif) [inline guide.md]: Wait are you saying "We can run our new migration", or are you saying you think we should rename the CLI command `diesel migration run`?

File: guide.md [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+new one, soon.
+
+Saved draft Diesel demo with id 1
+```
+
+Unfortunately, running `show_posts` still won't display our new post, because we
+saved it as a draft. We need to publish it! But in order to do that, we'll need
+to look at how to update an existing record. First, let's commit. The code for
+this demo at this point can be found [here][commit-no-2].
+
```
Comment (tessgriffin) [inline guide.md]: So at this point, we've saved a "draft"? So is it not actually written to the database yet? The next section we create the update functionality. So does it only get saved to Postgresql at that point?

Comment (sgrif) [inline guide.md]: It's in the database with `published = false`. Maybe I should expand on that more?

Comment (tessgriffin) [inline guide.md]: Yeah, I think so. I can't tell if the published part is cheeky because you're working with posts, or if that's an actual thing? I'm just confused as to where this is actually being saved to the database

Comment (sgrif) [inline guide.md]: I guess I assumed that it was implied that "drafts" are still saved somewhere

Comment (tessgriffin) [inline guide.md]: You should expand on it more. Is a "draft" something on the diesel side? At what exact point does the record get inserted into the database?

Comment (sgrif) [inline guide.md]: It's just a word for a post that isn't published (e.g. `published` is set to `false` in the database.) I set up the example this way to demonstrate that

a. Diesel doesn't fuck with your database defaults (we defaulted `published` to `false` in the migration)
b. To give me a bullshit reason to demonstrate how `update` works.

Clearly this is too implicit, though.

Comment (tessgriffin) [inline guide.md]: I think the terms are just confusing. You say show posts won't render because it's not 'published'. To me that implies that it's not saved in the database? I guess I don't relate 'publishing' to calling update. I might be clearer to have say a 'rough draft' when it's first saved into the database. And you're updating the post to be a 'final draft'

Comment (tessgriffin) [inline guide.md]: That could be partially my fault to. I didn't even realized there was a published column on the database. I think this example would probably be fine as long as you say "hey, we set this to be false by default, now let's update it to be true". That would make sense if it's more explicit.

Comment (sgrif) [inline guide.md]: Maybe I need to better call out the `published.eq(true)` section of https://github.com/sgrif/diesel/pull/181/files#diff-8c27d0d0b409c1de1867fdc9a128c72cR158 ?

Comment (sgrif) [inline guide.md]: FWIW I'm not trying to be confrontational or anything, just trying to identify *where* I need to improve things.

Comment (tessgriffin) [inline guide.md]: Yeah, that'd be a great thing to point out. And then to point out in your update that that's the column that you're changing.

File: guide.md [hunk 15]
```diff
@@ -0,0 +1,388 @@
...
+You know, a CLI application probably isn't the best interface for a blog demo.
+But really I just wanted a semi-simple example, where I could focus on Diesel.
+I didn't want to get bogged down in some web framework here.
+Plus I don't really like the Rust web frameworks out there. We might make a
+new one, soon.
+
+Saved draft Diesel demo with id 1
+```
+
+Unfortunately, running `show_posts` still won't display our new post, because we
```
Comment (tessgriffin) [inline guide.md]: I think even here on this line. You could say "Unfortunately, running `show_posts` still won't display our new post because we published to default to false in the database earlier. We'll have to 'update' our post to publish it!" or something like that.

File: guide.md (hunk 12 above)
Comment (derekprior) [inline guide.md]: I think the use of the term `run`, both in the cli and the documentation bothers me because regardless of the direction of the migration, you are `running` something.

Comment (sgrif) [inline guide.md]: :runner:

File: guide.md (hunk 9 above)
Comment (mfpiccolo) [inline guide.md]: I think something along those lines could be sufficient but I agree with Derek that it should have at least a quick explanation.

File: guide.md [hunk 16]
```diff
@@ -0,0 +1,388 @@
...
+```
+
+This will create our database for us (if it didn't already exist), and create an
+empty migrations directory that we can use to manage our schema. More on that
+later. Let's add Diesel to our `Cargo.toml`. We'll also add `dotenv` so we can
+use that same `.env` file for our environment variables.
+
+```toml
+[dependencies]
+diesel = "^0.5.0"
```
Comment (steveklabnik) [inline guide.md]: `^` is already the default with Cargo, so this and the others should just be "0.5.0"

File: guide.md [hunk 17]
```diff
@@ -0,0 +1,388 @@
...
+  body TEXT NOT NULL,
+  published BOOLEAN NOT NULL DEFAULT 'f'
+)
+
+-- down.sql
+DROP TABLE posts
+```
+
+We can run our new migration with `diesel migration run`. It's usually a good
+idea to make sure that `down.sql` is correct. You can do a quick sanity check by
```
Comment (steveklabnik) [inline guide.md]: ... this raises the question, when is it a good idea not _not_ make sure something is correct?

**Comment by sgrif** [comment]
I've addressed the feedback given so far. Can anyone take another look?

File: guide.md [hunk 18]
```diff
@@ -0,0 +1,406 @@
+Getting Started 2.0
+===
+
+For this guide, we're going to walk through some simple examples for each of the
+pieces of CRUD. Each step in this guide will build on the previous, and is meant
```
Comment (samphippen) [inline guide.md]: Is it worth explaining the acronym `CRUD` here. Generally, in guides like this, it's a good idea to define each term the first time it's used.

File: guide.md [hunk 19]
```diff
@@ -0,0 +1,406 @@
...
+cargo new diesel_demo
+cd diesel_demo
+```
+
+First, let's add Diesel to our dependencies. We're also going to use a tool
+called [`.env`][dotenv-rust] to manage our environment variables for us. We'll
+add it to our dependencies as well.
+
+[dotenv-rust]: https://github.com/slapresta/rust-dotenv
+
```
Comment (samphippen) [inline guide.md]: is it worth adding a thing that points out this is cargo.toml here?

Comment (sgrif) [inline guide.md]: Yes, this will go into those little terminal windows with the file name in the top bar

File: guide.md [hunk 20]
```diff
@@ -0,0 +1,406 @@
...
+echo DATABASE_URL=postgres://localhost/diesel_demo > .env
+```
+
+Now Diesel CLI can set everything up for us.
+
+```shell
+diesel setup
+```
+
+This will create our database for us (if it didn't already exist), and create an
```
Comment (samphippen) [inline guide.md]: the `for us` here is redundant.

File: guide.md [hunk 21]
```diff
@@ -0,0 +1,406 @@
...
+```
+
+You should see output which looks something like this:
+
+```
+Creating migrations/20160202154039_create_posts/up.sql
+Creating migrations/20160202154039_create_posts/down.sql
+```
+
+Migrations allow us to evolve the database schema over time. Each migration can
```
Comment (samphippen) [inline guide.md]: this is a beautiful sentence.

File: guide.md [hunk 22]
```diff
@@ -0,0 +1,406 @@
...
+be applied (`up.sql`) or reverted (`down.sql`). Applying and immediately
```
Comment (samphippen) [inline guide.md]: Applied and reverted are not antonyms. I wonder if there are two words here we could use that are true antonyms of each other.

File: guide.md (hunk 19 above)
Comment (sgrif) [inline guide.md]: screen shot 2016-02-03 at 10 48 38 am

File: guide.md [hunk 23]
```diff
@@ -0,0 +1,406 @@
...
+-- down.sql
+DROP TABLE posts
+```
+
+We can apply our new migration with `diesel migration run`. It's a good idea to
+make sure that `down.sql` is correct. You can do a quick sanity check by doing
+`diesel migration redo`, which will revert and reapply the latest migration.
+
+OK enough SQL, let's write some Rust. We'll start by writing some code to show
+us the last 5 published posts. The first thing we need to do is establish a
```
Comment (samphippen) [inline guide.md]: I might delete the "us" here.

File: guide.md [hunk 24]
```diff
@@ -0,0 +1,406 @@
...
+    dotenv().ok();
+
+    let database_url = env::var("DATABASE_URL")
+        .expect("DATABASE_URL must be set");
+    PgConnection::establish(&database_url)
+        .expect(&format!("Error connecting to {}", database_url))
+}
+```
+
+We'll also want to create a `Post` struct that we can read our data into, and
```
Comment (samphippen) [inline guide.md]: The dangling preposition on this clause makes me unhappy, can this read "into which we can read our data,"?

File: guide.md [hunk 25]
```diff
@@ -0,0 +1,406 @@
...
+have diesel generate the names we'll use to reference our tables and columns in
```
Comment (samphippen) [inline guide.md]: I would drop the "our" from "our tables and columns"

File: guide.md [hunk 26]
```diff
@@ -0,0 +1,406 @@
...
+#![feature(custom_derive, custom_attribute, plugin)]
+#![plugin(diesel_codegen, dotenv_macros)]
+
+pub mod schema;
+pub mod models;
+```
+
+The first two lines tell Rust that we want to use some special compiler plugins
+provided by Diesel and Dotenv. These will add various useful class annotations
+that we can use, as well as the `infer_schema!` macro which we'll see in just a
```
Comment (samphippen) [inline guide.md]: this sentence is very complex, is there any way of breaking it apart.

File: guide.md [hunk 27]
```diff
@@ -0,0 +1,406 @@
...
+    pub published: bool,
+}
+
+// src/schema.rs
+infer_schema!(dotenv!("DATABASE_URL"));
+```
+
+The `#[derive(Queryable)]` will generate all of the code needed to load a `Post`
+struct from a SQL query. The `infer_schema!` macro connects to the database URL
+given to it, and creates a bunch of code based on the database schema to
```
Comment (samphippen) [inline guide.md]: "The `infer_schema!` macro connects to the database URL which it is given. The macro also creates a bunch of code based on the database schema to represent all of the tables and columns."

File: guide.md (hunk 7 above)
Comment (sgrif) [inline guide.md]: screen shot 2016-02-03 at 10 53 45 am

File: guide.md [hunk 28]
```diff
@@ -0,0 +1,406 @@
...
+The `use posts::dsl::*` line imports a bunch of aliases so that we can say
+`posts` instead of `posts::table`, and `published` instead of
+`posts::published`. It's useful when we're only dealing with a single table, but
+that's not always what we want.
+
+We can run our script with `cargo run --bin show_posts`. Unfortunately, the results
+won't be terribly interesting, as we don't actually have any posts in the
+database. Still, we've written a decent amount of code, so let's commit.
+
+The full code for the demo at this point can be found [here][commit-no-1].
```
Comment (samphippen) [inline guide.md]: worth emboldening?

File: guide.md [hunk 29]
```diff
@@ -0,0 +1,406 @@
...
+----------
+
+You know, a CLI application probably isn't the best interface for a blog demo.
+But really I just wanted a semi-simple example, where I could focus on Diesel.
+I didn't want to get bogged down in some web framework here.
+Plus I don't really like the Rust web frameworks out there. We might make a
+new one, soon.
+```
+
+We've still only covered 3 of the 4 letters of CRUD though. Let's show how to
```
Comment (samphippen) [inline guide.md]: "let's show how to delete things" is an awkward sentence.

File: guide.md [hunk 30]
```diff
@@ -0,0 +1,406 @@
+Getting Started 2.0
+===
+
+For this guide, we're going to walk through some simple examples for each of the
+pieces of CRUD. Each step in this guide will build on the previous, and is meant
+to be followed along. This guide assumes that you're using PostgreSQL on
```
Comment (mcasper) [inline guide.md]: `is meant to be follow along.` feels a little awkward. Maybe reword to `is meant to be followed in order.`?

File: guide.md [hunk 31]
```diff
@@ -0,0 +1,406 @@
...
+
+The first thing we need to do is generate our project.
+
+```shell
+cargo new diesel_demo
+cd diesel_demo
+```
+
+First, let's add Diesel to our dependencies. We're also going to use a tool
+called [`.env`][dotenv-rust] to manage our environment variables for us. We'll
```
Comment (mcasper) [inline guide.md]: Nitpicking: The tool is called `dotenv`, should we list it as such? Especially since we have to put it in the toml as that

File: guide.md [hunk 32]
```diff
@@ -0,0 +1,406 @@
...
+
+Unfortunately, running `show_posts` still won't display our new post, because we
+saved it as a draft. If we look back to the code in `show_posts`, we added
+`.filter(published.eq(true))`, and we had `published` default to false in our
+migration. We need to publish it! But in order to do that, we'll need to look at
+how to update an existing record. First, let's commit. The code for this demo at
+this point can be found [here][commit-no-2].
+
+Now that we've got create and read out of the way, update is actually relatively
+simple. Let's jump write into the script:
```
Comment (mcasper) [inline guide.md]: `Let's jump right`

File: guide.md [hunk 33]
```diff
@@ -0,0 +1,406 @@
...
+    let connection = establish_connection();
+    let num_deleted = diesel::delete(posts.filter(title.like(pattern)))
+        .execute(&connection)
+        .expect("Error deleting posts");
+
+    println!("Deleted {} posts", num_deleted);
+}
+```
+
+We can run the script with `cargo run --bin delete_post demo` (at least with the
```
Comment (mcasper) [inline guide.md]: The paren on this line should probably be closed off on the next line, like
```
(at least with the title I chose. Your output should look something like):
```

Comment (sgrif) [inline guide.md]: I just like leaving unbalanced parens to make programmers feel uncomfortable (for the rest of the day

File: guide.md (hunk 22 above)
Comment (sgrif) [inline guide.md]: Open to suggestions.

File: guide.md (hunk 24 above)
Comment (sgrif) [inline guide.md]: This is the most British suggestion ever.

**Comment by sgrif** [comment]
Thanks everyone. I've addressed the last round of feedback in the implemented version. https://github.com/sgrif/diesel.rs-website/blob/master/source/views/guides/getting-started.html.slim
